### PR TITLE
Add FocusTrap & Keyboard closing to Modal

### PR DESCRIPTION
**Pull request by fspoettel**

### CODE REVIEW

File: src/components/Modal/Modal.js [hunk 1]
```diff
@@ -1,111 +1,113 @@
...
+    return (
+      <div>
+        <section
+          {...rest}
+          className={cx(
+            'slds-modal',
+            { [`slds-${transitionStyle}`]: open },
+            { 'slds-modal_prompt': !!prompt },
+            className
+          )}
```
Comment (philippfrank) [inline src/components/Modal/Modal.js]: Please extract the classes into a const.

File: src/components/Modal/Modal.js [hunk 2]
```diff
@@ -1,111 +1,113 @@
...
+  transitionStyle: PropTypes.oneOf([
+    'fade-in-open',
+    'slide-up-saving',
+    'slide-up-open',
+    'slide-down-cancel'
+  ]),
   children: PropTypes.node.isRequired,
-  /**
-   * class name
-   */
```
Comment (philippfrank) [inline src/components/Modal/Modal.js]: Why did you remove the comments?

File: src/components/Modal/Modal.js [hunk 3]
```diff
@@ -1,111 +1,113 @@
...
+    const name = child.type.displayName || child.type.name;
 
-    if (childName === 'ModalHeader') {
-      return React.cloneElement(child, {
-        key: label,
-        label,
-        prompt,
-        uncloseable: child.props.uncloseable !== undefined ? child.props.uncloseable : !!prompt,
-      });
+    if (name === 'ModalContent') {
```
Comment (philippfrank) [inline src/components/Modal/Modal.js]: Is something like `if (child.type === ModalHeader) { ... }` possible (without the hardcoded component name)?

File: src/components/Modal/Modal.js (hunk 2 above)
Comment (fspoettel) [reply src/components/Modal/Modal.js]: still need to add those again

File: src/components/Modal/Modal.js (hunk 3 above)
Comment (fspoettel) [reply src/components/Modal/Modal.js]: Sorta, but checking for a string is more performant here

Comment (philippfrank) [reply src/components/Modal/Modal.js]: Why?

Comment (fspoettel) [reply src/components/Modal/Modal.js]: Because a React Component is an object and shallow comparing objects of unknown shape is more expensive than comparing two strings.

Comment (philippfrank) [reply src/components/Modal/Modal.js]: Alright, I thought that `ModalHeader` would be a function and that `child.type` would be exactly that function, but you're probably right ✅
